### ISSUE TITLE
Using Workload API, unable to create SAS token on IoT Edge 1.1, works on 1.2

### ISSUE THREAD

**Issue by ksaye** [Azure/iotedge#5350]
## Expected Behavior
Following the documentation: https://github.com/Azure/iotedge/blob/master/edgelet/workload/docs/WorkloadApi.md, I am able to sign the URI and create a valid SAS Token on Edge 1.2, but i can not create a valid token on 1.1.

## Current Behavior
Using Azure IoT Explorer and the code below, I can create a matching SAS Token when on Edge 1.2.  This does not work on 1.1 and I have a custom module that needs to run on both versions of edge.

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. Setup IoT Edge 1.1 on Ubuntu 18.04 and deploy the module ksaye/ignition:8.1.4 (any module will work)
2. Using Azure IoT Explorer, generate a SAS Token, as shown below (token is expiration has passed for security/reporting reasons)
![image](https://user-images.githubusercontent.com/15837044/128934516-1a98f1a2-b3da-41eb-b333-b9d23331a67f.png)

3. Run the command ```docker run -i -t ignition bash``` to enter the container
4. Run the following command to get a SAS Token and **note it does not match**:
```
# expecting curl, base64 and jq to be installed in the Linux OS
epoch=1628629479
WORKLOADURI=$(echo $IOTEDGE_WORKLOADURI | sed "s|unix://|""|g")
uri=$(echo $IOTEDGE_IOTHUBHOSTNAME"/devices/"$IOTEDGE_DEVICEID"/modules/"$IOTEDGE_MODULEID | sed "s|/|%2F|g")
signedData=$(echo -n $uri$'\n'$epoch | base64 -w 0)
signature=$(curl -s --unix-socket $WORKLOADURI http://127.0.0.1/modules/$IOTEDGE_MODULEID/genid/$IOTEDGE_MODULEGENERATIONID/sign?api-version=$IOTEDGE_APIVERSION --request POST --data '{"data": "'$signedData'", "keyId": "", "algo": "HMAC-SHA256"}' | jq -r ".digest" | sed "s|=|%3D|g" | sed "s|+|%2B|g" | sed "s|/|%2F|g")
# here is our SAS Token
SASToken="SharedAccessSignature=SharedAccessSignature sr=$uri&sig=$signature&se=$epoch"
echo $SASToken
```

![image](https://user-images.githubusercontent.com/15837044/128934895-92dbb57a-8ea5-4199-8bdd-37a24ea2b42c.png)

5. Install IoT Edge 1.2 on Ubuntu 18.04 and deploy the module ksaye/ignition:8.1.4 (any module will work)
6. Using Azure IoT Explorer, generate a SAS Token, as shown below (token is expiration has passed for security/reporting reasons)

![image](https://user-images.githubusercontent.com/15837044/128935217-46187585-6b73-4131-a460-38f0a7708662.png)

7. Run the command ```docker run -i -t ignition bash``` to enter the container
8. Run the following command to get a SAS Token (the only change is epoch to match IoT Explorer) **note it does match**:
```
# expecting curl, base64 and jq to be installed in the Linux OS
epoch=1628629957
WORKLOADURI=$(echo $IOTEDGE_WORKLOADURI | sed "s|unix://|""|g")
uri=$(echo $IOTEDGE_IOTHUBHOSTNAME"/devices/"$IOTEDGE_DEVICEID"/modules/"$IOTEDGE_MODULEID | sed "s|/|%2F|g")
signedData=$(echo -n $uri$'\n'$epoch | base64 -w 0)
signature=$(curl -s --unix-socket $WORKLOADURI http://127.0.0.1/modules/$IOTEDGE_MODULEID/genid/$IOTEDGE_MODULEGENERATIONID/sign?api-version=$IOTEDGE_APIVERSION --request POST --data '{"data": "'$signedData'", "keyId": "", "algo": "HMAC-SHA256"}' | jq -r ".digest" | sed "s|=|%3D|g" | sed "s|+|%2B|g" | sed "s|/|%2F|g")
# here is our SAS Token
SASToken="SharedAccessSignature=SharedAccessSignature sr=$uri&sig=$signature&se=$epoch"
echo $SASToken
```

![image](https://user-images.githubusercontent.com/15837044/128935492-27d5072c-a9c8-41aa-b25f-bf88ba7e45a4.png)

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```

<Paste here between the triple backticks>

```
</details>

### Device Information
* Host OS [e.g. Ubuntu 18.04, Windows Server IoT 2019]:  Ubuntu 18.04
* Architecture [e.g. amd64, arm32, arm64]:  amd64
* Container OS [e.g. Linux containers, Windows containers]: Linux

### Runtime Versions
* aziot-edged [run `iotedge version`]: 1.1.4 and 1.2.3
* Edge Agent [image tag (e.g. 1.0.0)]: N/A
* Edge Hub [image tag (e.g. 1.0.0)]: N/A
* Docker/Moby [run `docker version`]: 20.10.7+azure

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs

<details>
<summary>aziot-edged logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-agent logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-hub logs</summary>

```

<Paste here between the triple backticks>

```
</details>

## Additional Information
Please provide any additional information that may be helpful in understanding the issue.

**Comment by nlcamp** [comment]
@ksaye - I'd like to use one of the following scripts to generate the SAS token so we can be confident that the SAS generated by the Azure IoT Explorer is correct: https://docs.microsoft.com/en-us/rest/api/eventhub/generate-sas-token

What are you setting the following variables to:

- $IOTEDGE_MODULEGENERATIONID
- $IOTEDGE_MODULEID
- $IOTEDGE_APIVERSION

**Comment by nlcamp** [comment]
@ksaye - as we've discussed offline, I was able to repro this issue. I'm just documenting this here while we work towards a fix.

**Comment by nlcamp** [comment]
Hi @ksaye - So we have a workaround for you that's pretty straightforward. Please specify "primary" as the "keyId" in the JSON value you send to the Workload API. In other words, change this:
`signature=$(curl -s --unix-socket $WORKLOADURI http://127.0.0.1/modules/$IOTEDGE_MODULEID/genid/$IOTEDGE_MODULEGENERATIONID/sign?api-version=$IOTEDGE_APIVERSION --request POST --data '{"data": "'$signedData'", "keyId": "", "algo": "HMAC-SHA256"}' | jq -r ".digest" | sed "s|=|%3D|g" | sed "s|+|%2B|g" | sed "s|/|%2F|g")`
to this:
`signature=$(curl -s --unix-socket $WORKLOADURI http://127.0.0.1/modules/$IOTEDGE_MODULEID/genid/$IOTEDGE_MODULEGENERATIONID/sign?api-version=$IOTEDGE_APIVERSION --request POST --data '{"data": "'$signedData'", "keyId": "primary", "algo": "HMAC-SHA256"}' | jq -r ".digest" | sed "s|=|%3D|g" | sed "s|+|%2B|g" | sed "s|/|%2F|g")`

With this change to your script, you should now be able to generate a SAS token that matches Azure IoT Explorer.

We're looking into whether we need to patch our code to default to "primary" in 1.1 or change the 1.2 behavior to be consistent with 1.1, but until then, this tweak to your script should unblock you.

Please let us know if you have any other questions.

**Comment by ksaye** [comment]
SWEET!  Let me run a quick test and them I will close the ticket.

**Comment by ksaye** [comment]
@nlcamp thanks for the solution, that does in fact work.

**Comment by nlcamp** [comment]
No problem @ksaye. Glad to see the workaround works for you. Just as an FYI, we consider this a bug that we plan to fix against 1.2. Thanks for reporting it!
